### PR TITLE
crimson/osd: record op's outdata and rval in pg log

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/pg.cc [hunk 1]
```diff
@@ -500,7 +502,15 @@ seastar::future<> PG::submit_transaction(ObjectContextRef&& obc,
 		      pg_log_entry_t::MODIFY : pg_log_entry_t::DELETE,
 		    obc->obs.oi.soid, osd_op_p.at_version, obc->obs.oi.version,
 		    osd_op_p.user_modify ? osd_op_p.at_version.version : 0,
-		    osd_op_p.req->get_reqid(), osd_op_p.req->get_mtime(), 0);
+		    osd_op_p.req->get_reqid(), osd_op_p.req->get_mtime(),
+                    op_info.allows_returnvec() && !ops.empty() ? ops.back().rval.code : 0);
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc:506]: how about moving this statement out, and returning the retval as a `future<int>`? so we don't need to repeat exactly the same code to calculate the retval of an op in `PG::do_osd_ops()`? something like
```c++
int result = 0;
if (op_info.allows_returnvec() && !ops.empty()) {
  result = ops.back();
}
log_entries.emplace_back( ///...
                      result,
                      /**/);
peering_state.append_log_with_trim_to_updated(/**/);
return backend->mutate_object(/**/).then(
  [this, result, /**/](/**/) {
  // ...
  return seastar::make_ready_future<int>(result);
}
```

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
```
03:54:24 - DEBUG    - cbt      - Nodes : incerta04.front.sepia.ceph.com
03:54:24 - DEBUG    - cbt      - Nodes : incerta04.front.sepia.ceph.com
03:54:24 - DEBUG    - cbt      - Nodes : incerta04.front.sepia.ceph.com
03:54:24 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 22.0918/27.9645  => rejected
03:54:24 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 5655.0/7158.0  => rejected
03:54:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 230.211/267.988  => accepted
03:54:24 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00280301/0.00222789  => rejected
03:54:24 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 22.3882/27.931  => rejected
03:54:24 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 5731.0/7150.0  => rejected
03:54:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 183.736/280.77  => accepted
03:54:24 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00275453/0.00222804  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 149.65/231.681  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 38310.0/59310.0  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 1931.84/1052.21  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000413987/0.000266753  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 150.166/226.869  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 38442.0/58078.0  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 1876.47/1341.37  => rejected
03:54:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000412637/0.000272672  => rejected
03:54:24 - WARNING  - cbt      - 14 tests failed out of 16
```

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
```
04:42:46 - DEBUG    - cbt      - Nodes : incerta02.front.sepia.ceph.com
04:42:46 - DEBUG    - cbt      - Nodes : incerta02.front.sepia.ceph.com
04:42:46 - DEBUG    - cbt      - Nodes : incerta02.front.sepia.ceph.com
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 22.4567/24.509  => rejected
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 5748.0/6274.0  => rejected
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 374.983/148.987  => rejected
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00276687/0.00252265  => rejected
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 22.1494/24.2869  => rejected
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 5670.0/6217.0  => rejected
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 405.699/182.511  => rejected
04:42:46 - WARNING  - cbt      - prefill/incerta02.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00276926/0.00255412  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 151.397/164.919  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 38757.0/42219.0  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 1772.93/1402.08  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000409617/0.000375707  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 150.272/167.353  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 38469.0/42842.0  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 2020.85/1401.64  => rejected
04:42:46 - WARNING  - cbt      - rand/incerta02.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000412707/0.00037035  => rejected
04:42:46 - WARNING  - cbt      - 16 tests failed out of 16
```

**Comment by tchaikov** [comment]
jenkins test crimson perf

File: src/crimson/osd/pg.cc (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/osd/pg.cc:506]: Would you be fine with addressing that with a dedicated commit?
I think a bigger refactor is needed and the `TODO` marker is already there:

```cpp
// TODO: refactor the submit_transaction
```

**Comment by rzarzynski** [comment]
```
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 24.2724/30.0305  => rejected
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 6213.0/7687.0  => rejected
15:08:14 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 244.924/299.467  => accepted
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00255954/0.00206885  => rejected
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 24.178/29.9031  => rejected
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 6189.0/7655.0  => rejected
15:08:14 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 256.952/272.512  => accepted
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.0025643/0.00207492  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 147.548/255.242  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 37772.0/65341.0  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 1479.64/471.698  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000420397/0.000241944  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 149.671/236.139  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 38315.0/60451.0  => rejected
15:08:14 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 1406.62/1176.65  => accepted
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000414366/0.000261584  => rejected
```

**Comment by rzarzynski** [comment]
jenkins retest perf

**Comment by tchaikov** [comment]
```
15:08:14 - DEBUG    - cbt      - Nodes : incerta04.front.sepia.ceph.com
15:08:14 - DEBUG    - cbt      - Nodes : incerta04.front.sepia.ceph.com
15:08:14 - DEBUG    - cbt      - Nodes : incerta04.front.sepia.ceph.com
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 24.2724/30.0305  => rejected
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 6213.0/7687.0  => rejected
15:08:14 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 244.924/299.467  => accepted
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00255954/0.00206885  => rejected
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 24.178/29.9031  => rejected
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 6189.0/7655.0  => rejected
15:08:14 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 256.952/272.512  => accepted
15:08:14 - WARNING  - cbt      - prefill/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.0025643/0.00207492  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 147.548/255.242  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 37772.0/65341.0  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 1479.64/471.698  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000420397/0.000241944  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 149.671/236.139  => rejected
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 38315.0/60451.0  => rejected
15:08:14 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 1406.62/1176.65  => accepted
15:08:14 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000414366/0.000261584  => rejected
15:08:14 - WARNING  - cbt      - 13 tests failed out of 16
```

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
```
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 28.7818/30.0305  => accepted
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 7368.0/7687.0  => accepted
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 178.72/299.467  => accepted
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00216118/0.00206885  => accepted
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 28.6254/29.9031  => accepted
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 7328.0/7655.0  => accepted
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 166.135/272.512  => accepted
15:31:24 - INFO     - cbt      - prefill/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00216677/0.00207492  => accepted
15:31:24 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 256.361/255.242  => accepted
15:31:24 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 65628.0/65341.0  => accepted
15:31:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.20)):: 992.406/471.698  => rejected
15:31:24 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000240725/0.000241944  => accepted
15:31:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 223.705/236.139  => rejected
15:31:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 57268.0/60451.0  => rejected
15:31:24 - INFO     - cbt      - rand/incerta04.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.20)):: 1114.28/1176.65  => accepted
15:31:24 - WARNING  - cbt      - rand/incerta04.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000276243/0.000261584  => rejected
15:31:24 - WARNING  - cbt      - 4 tests failed out of 16
```

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by rzarzynski** [comment]
```
CMake Error at src/CMakeLists.txt:210 (file):
  file STRINGS file
  "/home/jenkins-build/build/workspace/ceph-perf-pull-requests/ceph-master/src/.git_version"
  cannot be read.
```

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Pull request merged**